### PR TITLE
[WIP] Unconditionally admit already running pods in the TopologyManager after kubelet restart

### PR DESCRIPTION
**Pull request by klueska**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR fixes a (potential) bug in the `TopologyManager`, whereby *already-running* pods could be rejected after a kubelet restart if a device plugin that governs one of its resources had not reregistered with the kubelet yet.

Reference: https://github.com/kubernetes/kubernetes/issues/84479#issuecomment-552034118

An attempt to fix this issue another way was proposed here: https://github.com/kubernetes/kubernetes/pull/85038. However, this approach won't work at present because we do not checkpoint any NUMA information about devices allocated to containers. The only way to obtain this information is to wait for the device plugin to come back online so we can query it. This is impossible, however, as the underlying cause of the issue is that the plugin has not yet come online when this calculation is being made.

Things to note: 
* Even with this fix, there is a chance that a pod might be rejected if it was in the process of coming online as the kubelet restarted and a plugin has not yet reregistered (even if it normally would have been admitted and the containers that consume the plugin resources have already been allocated).

* This change will now allow pods to continue running even if the `TopologyManager` policy changes across a kubelet restart and the affinity properties of the pod are different than the ones allowed by the new policy. We need to decide if this is acceptable or not before proceeding with this PR.

**Does this PR introduce a user-facing change?**:

```release-note
After a kubelet restart, pods will now continue running even if the `TopologyManager` policy changes across a kubelet restart and the affinity properties of the pod are different than the ones allowed by the new policy.
```

### CODE REVIEW

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 1]
```diff
@@ -202,26 +202,37 @@ func (m *manager) RemoveContainer(containerID string) error {
 }
 
 func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
-	klog.Infof("[topologymanager] Topology Admit Handler")
+	// Short circuit to unconditionally admit the pod if the policy is 'none'.
 	if m.policy.Name() == "none" {
-		klog.Infof("[topologymanager] Skipping calculate topology affinity as policy: none")
```
Comment (cbf123) [inline pkg/kubelet/cm/topologymanager/topology_manager.go]: I kind of like this log as it represents an exit path from the function and we log the other exit paths.

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 2]
```diff
@@ -202,26 +202,37 @@ func (m *manager) RemoveContainer(containerID string) error {
...
 		}
 	}
+
 	pod := attrs.Pod
-	c := make(map[string]TopologyHint)
 
+	// Short circuit to unconditionally admit the pod if it is already running.
+	// This can happen after a kubelet restart, for example.
+	if pod.Status.Phase == v1.PodRunning && pod.DeletionTimestamp == nil {
+		klog.Infof("[topologymanager] Skipping ToplogyHint generation since pod already running")
```
Comment (cbf123) [inline pkg/kubelet/cm/topologymanager/topology_manager.go]: maybe add pod information to this log so we know what pod it applies to?

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 1 above)
Comment (lmdaly) [reply pkg/kubelet/cm/topologymanager/topology_manager.go]: +1

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 2 above)
Comment (lmdaly) [reply pkg/kubelet/cm/topologymanager/topology_manager.go]: at this stage it is guaranteed that all resources have been assigned to the pod right?

Comment (klueska) [reply pkg/kubelet/cm/topologymanager/topology_manager.go]: Yes. If it is already running, then there will be no more resources handed out to it.

**Comment by cbf123** [comment]
I've confirmed experimentally that this change resolves the issue I raised for the case where a device plugin is not yet registered with kubelet.

As for the question of whether it's okay to allow pods to keep running even if they don't meet the policy of the currently running kubelet, there's kind of precedence for this in the fact that changing the online CPUs of a node requires draining the node (according to https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/ ).  So maybe it wouldn't be totally crazy to say that changing the topology policy requires draining the node as well (to be totally robust, at least).

**Pull request closed without merging**